### ISSUE TITLE
Contextual attention in decoder, where Q != K

### ISSUE THREAD

**Issue by erickrf** [lucidrains/reformer-pytorch#23]
Hello @lucidrains and thanks for the nice implementation!
I have a question about using LSH in the contextual attention in a decoder, which can't have Q equal to K. 

In that case some queries might end up in buckets without keys, as the paper mentions, but there are ways around that - like picking the next best bucket (2nd or 3rd argmax) or just zeroing attention for that query. Do you think that could work?

In the context of MT, for example, if we can only use LSH in self-attention, then the encoder could even read some 64k tokens, but the decoder would need to be limited to ouput partial sequences (which sounds reasonable but not exploiting the full capabilities of the reformer).

**Comment by lucidrains** [comment]
@erickrf I think encoder / decoder with Reformer should work, although further efficiency could be explored. I am currently training a test example, outputting protein descriptions from amino acid sequences, and should know more by week's end. You could simply try it as well, following the example in the Readme. As for your questions, a masking operation takes place to prevent tokens from attending to itself. Your concern of missing a match is also taken care of by multiple rounds if hashing `n_hashes` which you can further tune as a hyperparameter.

**Comment by erickrf** [comment]
So, unless I misunderstood something in the code, it is not possible to use contextual attention with the current implementation, since it needs Q to be equal to K.

**Comment by lucidrains** [comment]
@erickrf I am uncertain how a shared QK space will behave for contextual attention as well, but I am currently projecting them into the same space https://github.com/lucidrains/reformer-pytorch/blob/master/reformer_pytorch/reformer_pytorch.py#L402 , experiments are still pending results.

**Comment by lucidrains** [comment]
@erickrf haha, just saw your edit. Yes, that's how I decided to approach it, because there's no way to guarantee equal amounts of Qs and Ks to be distributed per bucket to do anything more efficient. Perhaps a future paper will address that.

**Comment by lucidrains** [comment]
@erickrf this being a very empirical field, I would say, just try it! You can also adjust the `bucket_size` on the decoder to increase Q/K hitrates. Perhaps it will need to be adjusted higher than the default `64`

**Comment by lucidrains** [comment]
@erickrf please share your results too :) I think this is all uncharted territory
